### PR TITLE
Unify and simplify domains role state, improve state typing

### PR DESCRIPTION
**Pull request by JamesLefrere**
## Description

This PR makes a number of changes related to state, particularly re: domains and roles, in an attempt to make dealing with domain level permissions much simpler, with the idea that we ought to be able to use selectors to determine whether a user has a given role.

There are also significant typing improvements for state, so that `useSelector`/`useDataFetcher`/`useDataSubscriber` etc are all properly typed, arguments and return values! 🎉 

**New stuff** ✨

* Add a `roles` property to the `Domain` record
* Add selectors for domain roles
* Add a `ColonyRoles` enum (replacing colonyJS imports)
* Add a `fetchRoles` option to the `COLONY_DOMAINS_FETCH` action

**Changes** 🏗

* Go HAM on the typing of various state things such that the selector arguments and return values are typed for `useSelector`/`useDataFetcher`/`useDataSubscriber` etc 
* Unify the domains and roles state/queries where possible, such that users' roles are stored on the domains state
* Use the combined domain/roles state in various components, and make various simplifications along the way
* Convert the `domainId` field to a string, so that number->string coercing for this field does not cause unforeseen problems

**Deletions** ⚰️

* Remove the `allRoles` state and related hooks/fetchers/selectors
* Remove unused selectors (no point in typing all that!)

## TODO

- [x] Rebase!

### CODE REVIEW

**Comment by chmanie** [comment]
@ceolson01 @rdig can you please give this a review? It's not fully working yet but I don't want to make this much bigger. So I'll continue on this with #1879 and merge this into a feature branch that fixes everything

File: src/immutable/Using-Immutable.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
+# Using Immutable.js
```
[the remark this answers is not in the document]
Comment (chmanie) [reply src/immutable/Using-Immutable.md]: Credit goes to @JamesLefrere for this one!

**Comment by collinvine** [comment]
Nice work, guys 👍

**Comment by JamesLefrere** [comment]
Congrats on this one! 🎉

**Pull request merged**